txp/Pos/Txp/Toil/Types.hs

```diff
@@ -124,18 +127,41 @@ instance Default MemPool where
         }
 
 ----------------------------------------------------------------------------
--- ToilModifier
+-- UtxoModifier, UndoMap and AddrCoinsMap
 ----------------------------------------------------------------------------
 
 type UtxoModifier = MM.MapModifier TxIn TxOutAux
 type UndoMap = HashMap TxId TxUndo
-
-fromUtxo :: Utxo -> UtxoModifier
-fromUtxo = foldr (uncurry MM.insert) mempty . M.toList
+type AddrCoinMap = HashMap Address Coin
+
+utxoToModifier :: Utxo -> UtxoModifier
+utxoToModifier = foldr (uncurry MM.insert) mempty . M.toList
+
+-- | Takes utxo modifier and address-coin map with correspodning utxo
+-- and applies utxo modifier to map.
+-- Works for O(size of modifier * log (size of map)).
+applyUtxoModToAddrCoinMap
+    :: UtxoModifier
+    -> (AddrCoinMap, Utxo)
+    -> AddrCoinMap
+applyUtxoModToAddrCoinMap modifier (addrCoins, utxo) =
+    let outToPair = view _TxOut . toaOut in
+    let resolvedAddrs =
+          mapMaybe (fmap outToPair . flip M.lookup utxo)
+                   (MM.deletions modifier) in
+    let subAddress r c = if r < c then Just (c `unsafeSubCoin` r) else Nothing in
```

Comment:
Isn't that equivalent to just using `subCoin` instead of `unsafeSubCoin`?

Reply:
it's almost equivalent but I wouldn't like to keep zero balances